src/react/utils/useHydrateAtoms.ts

```diff
@@ -4,11 +4,22 @@ import type { WritableAtom } from '../../vanilla.ts'
 type Store = ReturnType<typeof useStore>
 type Options = Parameters<typeof useStore>[0]
 type AnyWritableAtom = WritableAtom<unknown, any[], any>
+type AtomTuple<A = AnyWritableAtom, V = unknown> = readonly [A, V]
+type InferAtoms<
+  T extends Array<AtomTuple>,
+  S extends Array<AtomTuple> = []
+> = S['length'] extends T['length']
+  ? S
+  : T extends Array<AtomTuple<infer A>>
+  ? A extends AnyWritableAtom
+    ? InferAtoms<T, [AtomTuple<A, ReturnType<A['read']>>, ...S]>
+    : T
+  : T
 
 const hydratedMap: WeakMap<Store, WeakSet<AnyWritableAtom>> = new WeakMap()
 
-export function useHydrateAtoms(
-  values: Iterable<readonly [AnyWritableAtom, unknown]>,
+export function useHydrateAtoms<T extends Array<AtomTuple>>(
```

Comment:
What do you think of using `Array` instead of `Iterable`?

Reply:
This would be a breaking change for `useHydrateAtoms`, using `Map` doesn't work with `Array`. Need to work on the `InferAtoms` some more.

I'm seeing some other `Map` errors when testing, but passing `Map` gives me the following:
![Screenshot 2023-04-27 at 11 22 39 AM](https://user-images.githubusercontent.com/525506/234927499-331af311-6d44-43c4-a0aa-4ef826b04443.png)

Reply:
Getting close using `Iterable`, but not perfect:
```ts
type InferIterableAtoms<T extends Iterable<AtomTuple>> = {
  [K in keyof T]: T[K] extends AtomTuple<infer A>
    ? A extends AnyWritableAtom
      ? AtomTuple<A, ReturnType<A['read']>>
      : T[K]
    : AtomTuple
}
export function useHydrateAtoms<T extends Iterable<AtomTuple>>(
  values: InferIterableAtoms<T>,
  options?: Options
) {
```
### Code
![Screenshot 2023-04-27 at 11 52 48 AM](https://user-images.githubusercontent.com/525506/234934471-0078d66d-f4f3-4696-9c54-a5cc0f279deb.png)

### Test
![Screenshot 2023-04-27 at 11 52 48 AM](https://user-images.githubusercontent.com/525506/234934419-ee637649-b20e-423b-8a00-51668f04b127.png)